### PR TITLE
Federated Clients: implement skeleton of federated consumer class

### PR DESCRIPTION
**Pull request by jonlee2**

### CODE REVIEW

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License").  See License in the project root for license information.
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java]: year is off

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaConsumerBuilder.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License").
 See License in the project root for license information.
+ */
+
+package com.linkedin.kafka.clients.consumer;
+
+import java.util.Map;
+
+
+public class LiKafkaConsumerBuilder<K, V> {
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaConsumerBuilder.java]: the default constructor uses a null map, which means new Builder().build() will get me an IllegalStateException. i think its more usable to default to an empty map.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
 See License in the project root for license information.
+ */
+
+package com.linkedin.kafka.clients.common;
+
+
+/**
+ * The common configurations for both LiKafkaProducer and LiKafkaConsumer
+ */
+public final class LiKafkaCommonClientConfigs {
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java]: i think our extensions should start with "lnkd" - so "lnkd.metadata.service.client". that way there's very little chance of a future collision, and its very clear which configs are extensions

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+      _mdsClient = mdsClient != null ? mdsClient :
+          configs.getConfiguredInstance(LiKafkaConsumerConfig.METADATA_SERVICE_CLIENT_CLASS_CONFIG,
+              MetadataServiceClient.class);
+
+      // Register this federated client with the metadata service. The metadata service will assign a UUID to this
+      // client, which will be used for later interaction between the metadata service and the client.
+      //
+      // Registration may also return further information such as the metadata server version and any protocol settings.
+      // We assume that such information will be kept and used by the metadata service client itself.
+      _clientId = mdsClient.registerFederatedClient(_clusterGroup, configs.originals());
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: probably needs a timeout (and that probably needs to be configurable)

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    } catch (Exception e) {
+      try {
+        if (_mdsClient != null) {
+          _mdsClient.close();
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: timeout?

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+        }
+      } catch (Exception e2) {
+        LOG.error("Failed to close the metadata service client", e2);
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: easier to just do 2.addSuppresed(e2) - would also cause the original exception to be printed 1st (more usable)

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java (hunk 1 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java]: done

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaConsumerBuilder.java (hunk 2 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaConsumerBuilder.java]: Done

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java (hunk 3 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java]: But I thought we want to be generic here? Like we can have a client for a non-LinkedIn metadata service?

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 6 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: Done

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 4 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: I added a common property for request timeout. We can expect each metadata client method would respect this timeout. Or did you actually want to specify timeout in each call?

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 5 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: See my comment above.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java (hunk 3 above)
Comment (radai-rosenblatt) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java]: the client can be an arbitrary implementation, but the property name that sets the client would start with "lnkd" so as not to be confused (or clash) with vanilla kafka client configs

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 4 above)
Comment (radai-rosenblatt) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: safer to always specify timeouts. clearer to read and less possibility of confusion

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java (hunk 3 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/common/LiKafkaCommonClientConfigs.java]: Misunderstood your original comment. As discussed in person, we'll use prefix "li." for new li-apache-kafka-clients specific property names.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 4 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: Added timeout to all mdsclient methods. Also added close() with maximum timeout for the use of AutoCloseable.

**Pull request merged**